File: C++/860. Lemonade Change.cpp
class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        int n = bills.size();
        int fives = 0,tens = 0,twens = 0;
        int indx = 0;
        for(auto ele:bills)
        {            
            if(ele == 5)fives += 1;
            else if(ele == 10)
            {
                tens += 1;
                if(fives>=1)fives -= 1;
                else return false;                
            }
            else if(ele == 20)
            {
                twens += 1;
                if(fives>=1 && tens>=1)fives -= 1,tens -= 1;
                else if(fives>=3 && tens == 0)fives -= 3;
                else return false;  
            }
        }
        return true;      
    }
};

File: JAVA/860. Lemonade Change.java
class Solution {
    int d_5 = 0;
    int d_10 = 0;
    public boolean lemonadeChange(int[] bills) {
        for (int i : bills) {
            if (i == 5) {
                d_5++;
            } else if (!change_available(i)) {
                return false;
            }
        }
        return true;
    }
    boolean change_available(int n) {
        if (n == 20) {
            if (d_5 >= 1 && d_10 >= 1) {
                d_5--;
                d_10--;
                return true;
            } else if (d_5 > 2) {
                d_5 -= 3;
                return true;
            }
            return false;
        }
        if (d_5 >= 1) {
            d_5--;
            d_10++;
            return true;
        }
        return false;
    }
}
